src/circuit/gadget/sinsemilla/chip/hash_to_point.rs

```diff
@@ -0,0 +1,446 @@
+                offset,
+                || y_a.ok_or(Error::SynthesisError),
+            )?;
+
+            // Assign lambda_2 and x_p zero values since they are queried
+            // in the gate.
+            {
+                region.assign_advice(
+                    || "dummy lambda2",
+                    config.lambda_2,
+                    offset,
+                    || Ok(pallas::Base::zero()),
+                )?;
+                region.assign_advice(
+                    || "dummy x_p",
+                    config.x_p,
+                    offset,
+                    || Ok(pallas::Base::zero()),
+                )?;
+            }
+
+            CellValue::new(y_a_cell, y_a.0)
+        };
+
+        #[cfg(test)]
+        #[allow(non_snake_case)]
+        // Check equivalence to result from primitives::sinsemilla::hash_to_point
+        {
+            use crate::circuit::gadget::sinsemilla::message::MessagePiece;
+            use crate::primitives::sinsemilla::{K, S_PERSONALIZATION};
+            use group::prime::PrimeCurveAffine;
+            use pasta_curves::arithmetic::CurveExt;
+
+            let field_elems: Option<Vec<pallas::Base>> =
+                message.iter().map(|piece| piece.field_elem()).collect();
+
+            if field_elems.is_some() {
+                // Get message as a bitstring.
+                let bitstring: Vec<bool> = message
+                    .iter()
+                    .map(|piece: &MessagePiece<pallas::Base, K>| {
+                        piece
+                            .field_elem()
+                            .unwrap()
+                            .to_le_bits()
+                            .into_iter()
+                            .take(K * piece.num_words())
+                            .collect::<Vec<_>>()
+                    })
+                    .flatten()
+                    .collect();
+
+                let hasher_S = pallas::Point::hash_to_curve(S_PERSONALIZATION);
+                let S = |chunk: &[bool]| hasher_S(&lebs2ip_k(chunk).to_le_bytes());
+
+                let expected_point = bitstring
+                    .chunks(K)
+                    .fold(Q.to_curve(), |acc, chunk| (acc + S(chunk)) + acc);
+                let actual_point =
+                    pallas::Affine::from_xy(x_a.value().unwrap(), y_a.value().unwrap()).unwrap();
+                assert_eq!(expected_point.to_affine(), actual_point);
+            }
+        }
+
+        Ok((EccPoint::from_coordinates_unchecked(x_a.0, y_a), zs_sum))
+    }
+
+    #[allow(clippy::type_complexity)]
+    // Hash a message piece containing `piece.length` number of `K`-bit words.
+    fn hash_piece(
+        &self,
+        region: &mut Region<'_, pallas::Base>,
+        offset: usize,
+        piece: &<Self as SinsemillaInstructions<
+            pallas::Affine,
+            { sinsemilla::K },
+            { sinsemilla::C },
+        >>::MessagePiece,
+        x_a: X<pallas::Base>,
+        y_a: Y<pallas::Base>,
```

Comment:
```suggestion
        y_a: Y<pallas::Base>,
        last_piece: bool,
```

Reply:
Use `final_piece`, to be consistent with comments and to distinguish from "last row".